Add codestral mamba2

molbap opened the pull request:
# What does this PR do?

This PR adds mamba2 to support mistral's mamba-codestral release.

- [x] Make sure weights are correctly loaded
- [x] Call correct cuda kernel in case of self.training, like mamba_inner_fn for mamba1
- [x] Cuda forward actually supports outputting states - we can use it for caching
- [x] Implement slow_forward as well
- [x] fix quality of slow forward generations
- [x] remove einops
- [x] Really remove all einops
- [x] Figure out why cuda kernels are actually slow? edit: it's unclear but an issue raised in the original repo https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306
- [x] add tests
- [x] Fix batched generation by tuning out states (thanks @vasqu !), edit: pointed out in https://github.com/state-spaces/mamba/issues/66#issuecomment-1863563829, affects mamba-1 as well likely
- [x] add more integration tests (batched generation)

cc @ArthurZucker

molbap committed:
add new model like

molbap committed:
draft cuda forward - mismatched keys (sharding on conv1)

molbap committed:
match keys successfully

molbap committed:
fix split

molbap committed:
get generation/forward running (wrong gens, norm?)

ArthurZucker committed:
:update

ArthurZucker committed:
some refactoring

ArthurZucker committed:
fixes

ArthurZucker committed:
works up until copy to cache

ArthurZucker committed:
fix

ArthurZucker committed:
update

ArthurZucker committed:
NON WORKING VERSION

ArthurZucker committed:
version that work?

ArthurZucker committed:
nit

molbap committed:
fix config

molbap committed:
fix conversion script

molbap committed:
working cuda forward

molbap committed:
fix merge conflict

ArthurZucker committed:
nit

ArthurZucker committed:
update

molbap committed:
Merge branch 'add_codestral_mamba2' of github.com:huggingface/new-model-addition-codestral-mamba2 into add_codestral_mamba2

ArthurZucker committed:
simplifcation

ArthurZucker committed:
make mamba slow simple work

ArthurZucker committed:
no einops

ArthurZucker committed:
todo

molbap committed:
fix style

ArthurZucker committed:
no einops

ArthurZucker committed:
update fix no einsum

ArthurZucker committed:
nit

molbap committed:
Merge branch 'add_codestral_mamba2' of github.com:huggingface/new-model-addition-codestral-mamba2 into add_codestral_mamba2

molbap committed:
Merge branch 'add_codestral_mamba2' of github.com:huggingface/new-model-addition-codestral-mamba2 into add_codestral_mamba2

molbap committed:
remove einops

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32080). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

vasqu commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,762 @@
+
+logger = logging.get_logger(__name__)
+
+if is_mamba_ssm_available():
```
Should this not be a version check since `mamba_ssm` only provides the necessary functions after v2.0 and there has been a bug on first release that caused NaN losses which has been fixed with [>=2.0.4](https://github.com/state-spaces/mamba/issues/352#issuecomment-2162223186). 

Further [info](https://github.com/state-spaces/mamba/issues/352#issuecomment-2167093395) that might be provided in the docs: There is a restriction on the parameters with `d_model * expand / headdim = multiple of 8`.

ArthurZucker replied:
cc @molbap we need to check version here!

vasqu reviewed (commented):
Hope you don't mind a few comments of mine. Good luck on the implementation, looking forward to the release!

molbap committed:
bug: scan_output differs strongly

Adibvafa commented:
I believe it will be more efficient to build on top of https://github.com/huggingface/transformers/pull/32027 to speed up the integration process!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,762 @@
+
+    # fmt: off
+    # FIXME slow generations are lower quality
+    def slow_forward(self, input_states, cache_params: Optional[Mamba2Cache]=None):
```
Yeah but the variant does not work for arbitrary length so we left it on the side for now!

ArthurZucker replied:
We'll checkout what you did!

ArthurZucker replied:
(no single letter variables pleasssssse 👿 )

vasqu replied:
Hard to do when it's referring to the SSM variables :sweat: the inner functions could probably be renamed instead of x I guess.

ArthurZucker commented:
Hey both! We tried your code and could not get it to work, mostly the shape of the output final_state is wrong (ends with 1024 vs 128 that we need). We are probably gonna ship it as is, but a patch will be possible if we can ship the ssimd in it!

vasqu commented:
@ArthurZucker I've just seen that it's using groups, so the code doesn't work out of the box as it would need appropriate resizing of B and C for example (ngroups -> nheads by repetition). I did not account for that as the original models never used anything other than group=1. Just tested it locally and it seems to work with those slight adjustments on B and C.  

I can't check in on the final_state shape without seeing the code so maybe I can help later when you guys ship it and there is a base to work on :D

molbap commented:
Hey both again, thanks for the comment @vasqu ! indeed it seems that it needs to be slightly modified - a `B.repeat(1, 1, self.num_heads // self.num_groups, 1)` seems promising. Are you verifying with a .generate() what the model is outputting? We'll ship this soon as Arthur mentioned and credit your code/include you as coauthors of course :hugs:

vasqu commented:
@molbap I've been using my [mamba2-torch](https://github.com/vasqu/mamba2-torch) repo to check modifications on the original `ssd_minimal` script to still work correctly (i.e. with variable seq_len, D skip connection, and just now with the n_groups >= 1). You can check it out over [here](https://github.com/vasqu/mamba2-torch/blob/main/tests/TestSSDMinimal.py). It does use einops though which I've changed in our PR to be pure torch.

molbap committed:
add rms norm option

molbap committed:
fix fast + slow generation with and w/o cache :heavy_check_mark:

molbap committed:
draft integration tests

molbap commented:
I changed up generations - the cuda kernel calls + the naive call have slightly different generations but both are very reasonable and seem similar to the original implementation. Missing some tests + remove the triple einsum contraction to create Ydiag and should be good to go!

vasqu commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1009 @@
+
+        # 3. Compute the inter-chunk SSM recurrence; produces correct SSM states at chunk boundaries
+        # (middle term of factorization of off-diag blocks; A terms)
+        if cache_params is not None and cache_params.seqlen_offset > 0:
+            previous_states = cache_params.ssm_states[self.layer_idx].unsqueeze(1)
```
Small suggestion on the generation via caching as we can avoid the following computations / simplify them just as in the fast path, see:
https://github.com/huggingface/transformers/blob/7e148149a1c809065e7ca720d170443c50bb7d69/src/transformers/models/mamba2/modeling_mamba2.py#L290-L327 

The equivalent for the slow path would be:
https://github.com/huggingface/transformers/blob/2e44073e4913b6ca26dbf7e0003cc07d36b0dd34/src/transformers/models/mamba2/modeling_mamba2.py#L1184-L1208

Maybe that helps even further in improving generation quality?

vasqu replied:
Mb, it needs to be adjusted for groups again. This should work (not thoroughly tested though):
```python
    # Get time step with softplus and bias
    # [bsz, 1, num_heads] -> [bsz, num_heads, head_dim]
    dt = dt.transpose(1, 2).expand(dt.shape[0], dt.shape[-1], self.head_dim)
    # [num_heads] -> [num_heads, head_dim]
    dt_bias = dt_bias.unsqueeze(-1).expand(dt_bias.shape[0], self.head_dim)
    dt = torch.nn.functional.softplus(dt + dt_bias.to(dt.dtype))
    
    
    # Discretize A
    # [num_heads] -> [num_heads, head_dim, state_size]
    A = A.unsqueeze(-1).unsqueeze(-1).expand(A.shape[0], self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
    # [bsz, num_heads, head_dim, state_size]
    dA = torch.exp(dt.unsqueeze(-1) * A)
    
    # Discretize B
    # [bsz, n_groups * state_size] -> [bsz, n_groups, 1, state_size] ->
    # -> [bsz, n_groups, group to head repetition factor, state_size] -> [bsz, num_heads, state_size]
    B = B.reshape(B.shape[0], self.n_groups, -1).unsqueeze(-2)
    B = B.expand(B.shape[0], B.shape[1], self.num_heads // self.n_groups, B.shape[-1]).contiguous()
    B = B.reshape(B.shape[0], -1, B.shape[-1])
    # [bsz, num_heads, head_dim, state_size]
    dB = dt.unsqueeze(-1) * B.unsqueeze(-2)
    
    # Discretize x into dB
    # [bsz, intermediate_size] -> [bsz, num_heads, head_dim]
    x = x.reshape(x.shape[0], -1, self.head_dim)
    dBx = dB * x.unsqueeze(-1)
    
    # State calculation
    cache_params.ssm_states[self.layer_idx].copy_(
        cache_params.ssm_states[self.layer_idx] * dA + dBx
    )
    
    # Subsequent output
    # [bsz, n_groups * state_size] -> [bsz, num_heads, state_size]
    C = C.reshape(C.shape[0], self.n_groups, -1).unsqueeze(-2)
    C = C.expand(C.shape[0], C.shape[1], self.num_heads // self.n_groups, C.shape[-1]).contiguous()
    C = C.reshape(C.shape[0], -1, C.shape[-1])
    # [bsz, num_heads, head_dim]
    y = torch.einsum("bhdn,bhn->bhd", cache_params.ssm_states[self.layer_idx].to(C.dtype), C)
    
    # D skip connection
    # [num_heads] -> [num_heads, head_dim]
    D = D.unsqueeze(-1).expand(D.shape[0], self.head_dim)
    y = y + x * D.unsqueeze(0)
    
    # [bsz, num_heads, head_dim] -> [bsz, 1, intermediate_size]
    y = y.reshape(y.shape[0], -1).unsqueeze(1)
```

molbap replied:
Yes, hadn't finished up the cached part entirely - I think your reshapes are correct, missing the dt clamping after softplus maybe?

vasqu replied:
Yes, totally forgot about that 👍

molbap committed:
remove a big chunk of the einsum

molbap committed:
fix slow, fast generations, without any einsum

molbap committed:
fix copies

molbap committed:
fix structure

molbap commented:
Alright, I've added a non-einsum equivalent of the tensor contraction paths for both the diag and off-diag terms that seem to give good results, supports batched inputs as well.

molbap commented:
@vasqu Haha, unfortunately it didn't it seems - there's still a difference in quality between the two paths. But I think it's reasonable for now, would need to do proper benchmarks.

vasqu commented:
Just saw the "Figure out why cuda kernels are actually slow?"

I've also wondered about that when i worked on mamba2 and here's an issue of mine I used to keep track: https://github.com/vasqu/mamba2-torch/issues/1#issue-2349175830

It seems like the triton kernels cause a significant cpu overhead on the first iteration(s): https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306

Other suggestions: big model and/or compiling https://github.com/state-spaces/mamba/issues/355#issuecomment-2147597457 (but that breaks in a lot of cases which happened in my case too)

molbap commented:
Yes! Been reading that one too and couldn't get the compile to improve performances (ie crashes). The warmup run definitely helps as well. So this full torch path is currently faster.

molbap committed:
fix up modeling and tests

molbap commented:
I added some tests and did last touches on modeling - Batched generation is a bit weird, I think there is leaking of states within a batch and not sure this is fixed in the original repo either so keeping it as is.

molbap committed:
fix tests

molbap committed:
Merge branch 'main' into add_codestral_mamba2

molbap committed:
clamping is indeed worse

molbap committed:
recover mamba2 cache test

molbap committed:
fix copies

molbap committed:
no cache position (yet)

molbap committed:
fix tf tests

molbap committed:
fix matmul for generate

molbap committed:
fixup

molbap committed:
skip cache tests for now

molbap committed:
[run-slow]mamba2

vasqu commented:
Late night me, regarding batched generation: Do the padding sides differ much (left vs right)? I'd assume that we have leaky information flowing through either way. This affects both mamba 1+2. 

When using left padding we would pad with the eos token and subsequently with a non-zero input which would influence causal directions (left to right, repeated eos information is given to the SSM and conv1d). In attention a simple attention mask would take care of it for the core softmax operations, for mamba there does not exist such thing yet. I'd suggest using the original attention mask to zero the padding tokens before entering the mamba block (zero input and zero initial hidden state would result in repeated zeros until encountering the first non zero input). It has been similarly suggested back then for the original mamba model: https://github.com/state-spaces/mamba/issues/66#issuecomment-1863563829

Right padding would probably need keeping track of the sequence indices which recent versions of mamba and conv1d support. Without recent versions it would probably get less "elegant".

molbap commented:
Indeed, that's likely the issue, thanks @vasqu ! I tried tuning out the hidden states but didn't think of doing it in two places :face_with_peeking_eye: It seems that fixes the batched generation, I'll push that soon

molbap committed:
tune out hidden states for padding

molbap committed:
test batched generation

vasqu commented:
@molbap Does that also work for right padding? Reassigning zeros would probably spit out garbage imo (without retrieving the correct last values). Maybe raise a warning and do not interact with the attention mask if right padding?

molbap committed:
propagate attention mask changes

molbap commented:
I don't think it does right now for right padding no, but it should work in all other cases now at least!

molbap committed:
fix past length

molbap committed:
fix integration test

molbap committed:
style

ArthurZucker commented on docs/source/en/model_doc/mamba2.md:
```diff
@@ -0,0 +1,50 @@
+
+## Overview
+
+The mamba2 model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
+<INSERT SHORT SUMMARY HERE>
+
+The abstract from the paper is the following:
+
+*<INSERT PAPER ABSTRACT HERE>*
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
+
+This model was contributed by [INSERT YOUR HF USERNAME HERE](https://huggingface.co/<INSERT YOUR HF USERNAME HERE>).
+The original code can be found [here](<INSERT LINK TO GITHUB REPO HERE>).
+
```
this needs filling!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+
+    Assumes that we only have tensors of either size 4 or 3
+    """
+    assert 2 < len(x.shape) < 5
```
no asserts please

molbap replied:
woops, missed it

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+        return x.reshape(x.shape[0], -1, chunk_size, x.shape[2], x.shape[3])
+
+
+def segsum(x):
```
let's avoid single letter variable (valid for all func)

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+    x = pad_by_size(x, pad_size)
+
+    if len(x.shape) == 3:
+        # b (l c) h -> b l c h with c=chunk_size
```
```suggestion
```

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+        # [bsz, seq_len multiple of chunk_size, num_heads] -> [bsz, -1, chunk_size, num_heads]
+        return x.reshape(x.shape[0], -1, chunk_size, x.shape[2])
+    else:
+        # b (l c) h p -> b l c h p with c=chunk_size
```
```suggestion
```

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+        if cache_params is not None and cache_params.seqlen_offset > 0:
+            batch_size = hidden_states.shape[0]
+            zxbcdt = self.in_proj(hidden_states.squeeze(1))  # (B 2D)
+            d_mlp = (
+                zxbcdt.shape[-1]
+                - 2 * self.intermediate_size
+                - 2 * self.n_groups * self.ssm_state_size
+                - self.num_heads
+            ) // 2
+
+            z0, x0, gate, xBC, dt = torch.split(
+                zxbcdt,
+                [
+                    d_mlp,
+                    d_mlp,
+                    self.intermediate_size,
+                    self.intermediate_size + 2 * self.n_groups * self.ssm_state_size,
+                    self.num_heads,
+                ],
+                dim=-1,
+            )
```
let's spslit these as much as possible to make it readable

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            )
+            A = -torch.exp(self.A_log.float())  # (nheads,)
+
+            A = A.unsqueeze(1).unsqueeze(2).expand(-1, self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
+            dt = dt.unsqueeze(2).expand(-1, -1, self.head_dim)
+            dt_bias = self.dt_bias.unsqueeze(1).expand(-1, self.head_dim)
+            D = self.D.unsqueeze(1).expand(-1, self.head_dim)  # repeat(self.D, "h -> h p", p=self.head_dim)
```
I am more a fan of [:,None,None] notation!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+                dt_softplus=True,
+            )
+            hidden_states = hidden_states.view(batch_size, self.num_heads * self.head_dim)
+            if self.rms_norm:
+                hidden_states = self.norm(hidden_states, gate)
```
I think rms norm is always used, let's remove if else

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            hidden_states = hidden_states.view(batch_size, self.num_heads * self.head_dim)
+            if self.rms_norm:
+                hidden_states = self.norm(hidden_states, gate)
+            if d_mlp > 0:
```
never 0 here let's remove

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+                    self.conv1d.bias,
+                    self.dt_bias,
+                    A,
+                    D=self.D.view(-1, self.head_dim) if self.D_has_hdim else self.D,
```
LEt's remove the if else here as well!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+                )
+                if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
+                    # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
+                    # bug in generate tests?
```
potentially because some test use 4d mask!

molbap replied:
I think I got rid of all generate bugs in the end, they were related to the attention mask slicing!

ArthurZucker replied:
cool!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+                    dim=-1,
+                )
+
+                if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
+                    # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
+                    dtype = hidden_states.dtype
+                    hidden_states = (hidden_states * attention_mask.unsqueeze(2)).to(dtype)
```
why are we doing it twice? Interesting

molbap replied:
it's because of the convolution: even if we tune out the hidden states before it, we need to tune the remainder again to 0 afterwards!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+
+            dt = torch.nn.functional.softplus(dt + dt_bias.to(dt.dtype))
+            dt = torch.clamp(dt, self.time_step_min) #, self.time_step_max)
+            A = A[..., None, None].expand(A.shape[0], self.head_dim, self.ssm_state_size).to(dtype=torch.float32)
```
shape[0] is batch size, reused a lot, let's define them, same for sjape[1] and shape[-1]

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+
+            # Subsequent output
+            # [bsz, n_groups * state_size] -> [bsz, num_heads, state_size]
+            C = C.reshape(C.shape[0], self.n_groups, -1).unsqueeze(-2)
+            C = C.expand(C.shape[0], C.shape[1], self.num_heads // self.n_groups, C.shape[-1]).contiguous()
+            C = C.reshape(C.shape[0], -1, C.shape[-1])
+            # [bsz, num_heads, head_dim]
```
explicit shape will make this more readable + None expansion

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            M = M_intermediate.sum(dim=-1)
+
+            # Step 3: Compute Y_diag (apply to values)
+            #Y_diag = ((M.unsqueeze(-1) * hidden_states.unsqueeze(1)).sum(dim=3))
+            #Y_diag_einsum = torch.einsum("bclsh,bcshp->bclhp", M, hidden_states)
+            # Y_diag_alt = (M.unsqueeze(-1) * hidden_states.unsqueeze(2)).sum(3)
+            #diff_ = ((Y_diag_alt - Y_diag_einsum) / (Y_diag_einsum + 1e-9)).min()
+
+            #Y_diag_intermediate = M[..., None] * hidden_states[:, None, ...]
+            #Y_diag = Y_diag_intermediate.sum(dim=3)
+
```
to remove?

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            #Y_diag_intermediate = M[..., None] * hidden_states[:, None, ...]
+            #Y_diag = Y_diag_intermediate.sum(dim=3)
+
+            Y_diag = (M.unsqueeze(-1) * hidden_states.unsqueeze(2)).sum(3)
```
None notationt!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            decay_chunk = torch.exp(segsum(nn.functional.pad(A_cumsum[:, :, :, -1], (1, 0))))
+
+            states_permuted = states.permute(0, 2, 1, 3, 4)
+            result = (decay_chunk[..., None, None] * states_permuted[:, :, None, ...]).sum(dim=2)
```
permuting decay chunk could avoid 2 permutes? (just wondering but ignore this comment!)

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            # [bsz, num_heads, head_dim] -> [bsz, 1, intermediate_size]
+            y = y.reshape(y.shape[0], -1).unsqueeze(1)
+        else:
+            # begin ssd naive implementation without einsums
```
wow that's heavy thanks for figuring it out!

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+    def set_input_embeddings(self, new_embeddings):
+        return self.backbone.set_input_embeddings(new_embeddings)
+
+    def _update_model_kwargs_for_generation(
+        self, outputs: ModelOutput, model_kwargs: Dict[str, Any], num_new_tokens: int = 1, **kwargs
+    ) -> Dict[str, Any]:
+        model_kwargs["cache_params"] = outputs.get("cache_params", None)
+        if (
+            model_kwargs.get("use_cache", True)
+            and "cache_position" in model_kwargs
+            and model_kwargs["cache_position"] is not None
+        ):
+            model_kwargs["cache_position"] = model_kwargs["cache_position"][-1:] + num_new_tokens
+
+        return model_kwargs
```
if you are not using them for now, ff to remove

ArthurZucker commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,1123 @@
+            if cache_position[0] > 0:
+                input_ids = input_ids[:, -1].unsqueeze(-1)
+                attention_mask = attention_mask[:, -1].unsqueeze(-1)
+            else:
+                # we initialize the `cache_position` to full size of `conv_states` at prefill stage
+                # considering padding will be applied when input length is shorter, and truncation
+                # will be applied when it is longer, so it will be equivalent to always have it match
+                # the length of `cache_params.conv_states`, which is `config.conv_kernel`
+                cache_position = torch.arange(0, past_len, device=input_ids.device)
+                # if the cache is not used, we also do have to extend the attention mask here
+                # TODO there is likely a cleverer way to do this
+                extended_mask = torch.ones(
+                    attention_mask.size(0), past_len - attention_mask.shape[1], device=attention_mask.device
+                )
+                attention_mask = torch.cat([attention_mask, extended_mask], dim=1)
+                cache_params = None
+
+        if attention_mask.shape[1] < past_len:
+            # we have to update manually the attention mask if
+            # we are in decoding without cache
+            # and we don't have position_ids here
+            # TODO but we should be able to use cache_position though at a later time
+            extended_mask = torch.ones(
+                attention_mask.size(0), past_len - attention_mask.shape[1], device=attention_mask.device
+            )
+            attention_mask = torch.cat([attention_mask, extended_mask], dim=1)
+        if inputs_embeds is not None and cache_params is None:
+            model_inputs = {"inputs_embeds": inputs_embeds}
+        else:
+            model_inputs = {"input_ids": input_ids}
+
+        model_inputs.update(
+            {
+                "attention_mask": attention_mask,
+                "cache_params": cache_params,
+                "use_cache": use_cache,
+                "cache_position": cache_position,
+            }
+        )
+        return model_inputs
```
we need this only if you use the cache positions

molbap replied:
We need it too for the attention_mask, right?

ArthurZucker reviewed (commented):
Congrats! There are a few nits here and there but LGTM otherwise!

ArthurZucker commented:
The `.md` file should include your findings about speed performances etc

molbap committed:
address comments

molbap committed:
update readme

molbap committed:
add mamba2 version check

molbap committed:
fix tests

molbap committed:
[run-slow]mamba2

molbap commented:
Addressed comments and fixed some more tests, also tried a simple PEFT script (almost same as mamba 1) which seems to run :grinning: added it to the .md.

molbap committed:
skip edge tests

molbap committed:
[run-slow]mamba2

molbap committed:
last fixup

molbap committed:
[run-slow]mamba2

molbap commented:
cc @ArthurZucker , the only test failing is the slow test because runner doesn't have access

vasqu commented on docs/source/en/model_doc/mamba2.md:
```diff
@@ -0,0 +1,97 @@
+Tips:
+
+This version should support all implementations of Mamba 2, and in particular [Mamba-2 codestral](https://huggingface.co/mistralai/Mamba-Codestral-7B-v0.1) from Mistral AI. In particular, mamba 2 codestral was released with a number of `groups` equal to 8, which can be thought intuitively as similar to the number of kv heads in an attention-based model. 
+This model has two different forward passes, `torch_forward` or `cuda_kernels_forward`. The latter uses the original cuda kernels if they are found in your environment, and is slower on the prefill i.e. requires a "warmup run" due to high cpu overhead, see [here](https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306) and [also here](https://github.com/state-spaces/mamba/issues/ ...
```
Small nit: Can we add left padding to the tips so people avoid using right padding (which likely doesn't work as expected) :eyes:

molbap replied:
It's by default activated, but you're right, better mention it here!

vasqu commented on src/transformers/models/mamba2/modeling_mamba2.py:
```diff
@@ -0,0 +1,762 @@
+            residual = residual.to(torch.float32)
+
+        hidden_states = self.mixer(hidden_states, cache_params=cache_params)
+        hidden_states = residual + hidden_states
```
I think I've had it wrong here since the norm takes care of the downcast... mb

It will lead to fp32 in the last output of the block though which might be avoided by a downcast. For example, a bf16 input is computed correctly but the last output given to the rms norm ("norm_f") will take in fp32 and subsequently gets fp32 again. This leads to dtype mismatches if we expect a bf16 output for other subsequent components. 

In short, calculations for fp16/bf16 are correct but we get fp32 as the final output which shouldn't be the case imo.

vasqu replied:
cc @molbap not sure about this one but would love to hear your opinion on this.

vasqu commented:
Looking really great! :fire: Thanks for the work

ArthurZucker commented on docs/source/en/model_doc/mamba2.md:
```diff
@@ -0,0 +1,97 @@
+    train_dataset=dataset,
+    dataset_text_field="quote",
+)
+trainer.train()
+## Mamba2Config
```
```suggestion
trainer.train()
```

## Mamba2Config
```

molbap replied:
ah, just fixed that in the last README update!

ArthurZucker reviewed (approved):
LGTM, let's fix CI and we can merge!

molbap committed:
update README

ArthurZucker commented on docs/source/en/model_doc/mamba2.md:
```diff
@@ -29,6 +29,7 @@ Tips:
 This model has two different forward passes, `torch_forward` or `cuda_kernels_forward`. The latter uses the original cuda kernels if they are found in your environment, and is slower on the prefill i.e. requires a "warmup run" due to high cpu overhead, see [here](https://github.com/state-spaces/mamba/issues/389#issuecomment-2171755306) and [also here](https://github.com/state-spaces/mamba/issues/ ...
+Another note, shutdown of hidden states corresponding to padding tokens is done in 2 places and mostly has been tested with left-padding. Right-padding will propagate noise down the line and is not guaranteed to yield satisfactory results. `tokenizer.padding_side = "left"` ensures you are using the correct padding side.
 
 This model was contributed by [Molbap](https://huggingface.co/Molbap), with tremendous help from [Anton Vlasjuk](https://github.com/vasqu).
```
cc @vasqu thanks for your input in this PR! 🥳

molbap replied:
yes thanks a lot @vasqu , glad to have had your insights along the way, it was incredibly helpful!

Adibvafa commented:
Super excited for this merge!
@vasqu Should we close https://github.com/huggingface/transformers/pull/32027 now?

vasqu commented:
@Adibvafa Sry, it kinda drowned in my notifications! I'll close it for now, but if needed, it can also be reopened at any time.
